### PR TITLE
[Reimbursements] Create Pending Transactions when Reimbursement is Requested

### PR DESCRIPTION
**Pull request by sampoder**
Closes https://github.com/hackclub/hcb/issues/5763. Am I doing this right @garyhtou?

### CODE REVIEW

File: app/services/reimbursement/expense_payout_service/nightly.rb [hunk 1]
```diff
@@ -19,15 +21,17 @@ def run
...
         end
 
         Reimbursement::ExpensePayout.pending.find_each(batch_size: 100) do |expense_payout|
+          next if !expense_payout.reimbursement_approved?
+
           Reimbursement::ExpensePayoutService::ProcessSingle.new(expense_payout_id: expense_payout.id).run
         end
 
         Reimbursement::PayoutHolding.pending.find_each(batch_size: 100) do |payout_holding|
+          next if !payout_holding.reimbursement_approved?
```
Comment (sampoder) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: Should be the report

File: app/services/reimbursement/expense_payout_service/nightly.rb [hunk 2]
```diff
@@ -19,15 +21,17 @@ def run
             report:
           )
 
-          report.mark_reimbursed!
-
         end
 
         Reimbursement::ExpensePayout.pending.find_each(batch_size: 100) do |expense_payout|
+          next if !expense_payout.reimbursement_approved?
```
Comment (sampoder) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: Should be the report

Comment (sampoder) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: Use unless

File: app/services/reimbursement/expense_payout_service/nightly.rb (hunk 1 above)
Comment (sampoder) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: Use unless

File: app/services/reimbursement/expense_payout_service/nightly.rb [hunk 3]
```diff
@@ -41,6 +45,7 @@ def run
           if payout_holding.canonical_transactions.any?
             payout_holding.mark_settled!
           end
+          payout_holding.report.mark_reimbursed!
```
Comment (garyhtou) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: maybe move it inside process single

File: app/services/reimbursement/expense_payout_service/nightly.rb (hunk 2 above)
Comment (garyhtou) [inline app/services/reimbursement/expense_payout_service/nightly.rb]: create a new state for "processing"

File: app/models/reimbursement/report.rb [hunk 4]
```diff
@@ -108,6 +108,11 @@ class Report < ApplicationRecord
         transitions from: [:draft, :submitted, :reimbursement_requested], to: :rejected
         after do
           ReimbursementMailer.with(report: self).rejected.deliver_later
+          if payout_holding.present?
+            payout_holding.expense_payouts.each do |expense_payout|
+              expense_payout.canonical_pending_transaction&.decline!
+            end
+          end
```
Comment (garyhtou) [inline app/models/reimbursement/report.rb]:
```suggestion
          payout_holding&.destroy!
```

```ruby
class PayoutHolding
  has_many :expense_payouts, dependent: :destroy
end

class ExpensePayout
  before_destroy do
    canonical_pending_transaction&.decline!
  end
end
```

**Review by garyhtou — COMMENTED** [review]
🪄

1. Disable job in prod
2. Merge
3. Fix production states
4. Enable job

**Pull request merged**